x/incentive/keeper/adapters/hard_borrow/adapter_test.go

```diff
@@ -0,0 +1,400 @@
+	hardKeeper := suite.app.GetHardKeeper()
+
+	err := hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(40_000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(25_000)),
+		),
+	)
+	suite.NoError(err)
+
+	adapter := hard_borrow.NewSourceAdapter(suite.app.GetHardKeeper())
+
+	tests := []struct {
+		name         string
+		giveSourceID string
+		wantShares   sdk.Dec
+	}{
+		{
+			"total shares",
+			suite.denomA,
+			sdk.NewDecWithPrec(40_000+25_000, 0),
+		},
+		{
+			"empty or invalid denom empty",
+			"denom2",
+			sdk.ZeroDec(),
+		},
+	}
+
+	for _, tt := range tests {
+		suite.Run(tt.name, func() {
+			shares := adapter.TotalSharesBySource(suite.ctx, tt.giveSourceID)
+
+			suite.Equal(tt.wantShares, shares)
+		})
+	}
+}
+
+func (suite *HardBorrowAdapterTestSuite) TestHardAdapter_TotalSharesBySource_Interest() {
+	hardKeeper := suite.app.GetHardKeeper()
+
+	err := hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(40_000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(25_000)),
+		),
+	)
+	suite.NoError(err)
+
+	// ------------------------------------------
+	// Accrue interest for hard borrows
+	suite.ctx = suite.ctx.WithBlockTime(suite.ctx.BlockTime().Add(time.Hour))
+	hard.BeginBlocker(suite.ctx, hardKeeper)
+	suite.ctx = suite.ctx.WithBlockTime(suite.ctx.BlockTime().Add(24 * time.Hour))
+	hard.BeginBlocker(suite.ctx, hardKeeper)
+
+	// ------------------------------------------
+	// Modify borrow to sync with new interest - Index field
+	// Second borrow has no modifications
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(1_000)),
+		),
+	)
+	suite.NoError(err)
+
+	adapter := hard_borrow.NewSourceAdapter(suite.app.GetHardKeeper())
+
+	tests := []struct {
+		name         string
+		giveSourceID string
+		wantShares   sdk.Dec
+	}{
+		{
+			"total shares",
+			suite.denomA,
+			// 40_000 + 25_000 + 1_000 + interest
+			// 66_000 + interest
+			sdk.NewDecWithPrec(66002, 0),
```

Comment:
We'll need to exclude interest from the total shares as well. Then the `normalized_owner_borrow / normalized_total_borrows` should be fraction of the borrows a user owns.
The old normalisation code: https://github.com/Kava-Labs/kava/blob/cb9e3da74aef12149a9c9a3746312aab304f46d2/x/incentive/keeper/rewards_borrow.go#L39-L64